icu4j/main/classes/core/src/com/ibm/icu/util/LocaleMatcher.java

```diff
@@ -54,880 +63,958 @@
+         */
+        NONE,
+        /**
+         * Earlier desired locales are preferred.
+         *
+         * <p>From each desired locale to the next,
+         * the distance to any supported locale is increased by an additional amount
+         * which is at least as large as most region mismatches.
+         * A later desired locale has to have a better match with some supported locale
+         * due to more than merely having the same region subtag.
+         *
+         * <p>For example: <code>Supported={en, sv}  desired=[en-GB, sv]</code>
+         * yields <code>Result(en-GB, en)</code> because
+         * with the demotion of sv its perfect match is no better than
+         * the region distance between the earlier desired locale en-GB and en=en-US.
+         *
+         * <p>Notes:
+         * <ul>
+         *   <li>In some cases, language and/or script differences can be as small as
+         *       the typical region difference. (Example: sr-Latn vs. sr-Cyrl)
+         *   <li>It is possible for certain region differences to be larger than usual,
+         *       and larger than the demotion.
+         *       (As of CLDR 35 there is no such case, but
+         *        this is possible in future versions of the data.)
+         * </ul>
+         *
+         * @draft ICU 65
+         * @provisional This API might change or be removed in a future release.
+         */
+        REGION
     }
 
     /**
-     * Get the best match for a LanguagePriorityList
+     * Data for the best-matching pair of a desired and a supported locale.
      *
-     * @param languageList list to match
-     * @return best matching language code
-     * @stable ICU 4.4
+     * @draft ICU 65
+     * @provisional This API might change or be removed in a future release.
      */
-    public ULocale getBestMatch(LocalePriorityList languageList) {
-        double bestWeight = 0;
-        ULocale bestTableMatch = null;
-        double penalty = 0;
-        OutputDouble matchWeight = new OutputDouble();
-        for (final ULocale language : languageList) {
-            final ULocale matchLocale = getBestMatchInternal(language, matchWeight);
-            final double weight = matchWeight.value * languageList.getWeight(language) - penalty;
-            if (weight > bestWeight) {
-                bestWeight = weight;
-                bestTableMatch = matchLocale;
-            }
-            penalty += 0.07000001;
-        }
-        if (bestWeight < threshold) {
-            bestTableMatch = defaultLanguage;
+    public static final class Result {
+        private final ULocale desiredULocale;
+        private final ULocale supportedULocale;
+        private final Locale desiredLocale;
+        private final Locale supportedLocale;
+        private final int desiredIndex;
+        private final int supportedIndex;
+
+        private Result(ULocale udesired, ULocale usupported,
+                Locale desired, Locale supported,
```

Comment:
Do we really need to store Locale in parallel to ULocale here?

Reply:
The constructor always makes supported Locales from the supported ULocales, so both versions just get their references copied here. If someone has a build-match-discard use case, we can revisit lazy-building one or the other version.

One of the desired objects is real (reference copied from the input), the other is null. Their getters convert on the fly -- but I expect that if you match one type you won't normally ask to get the other. I considered making the getters cache the conversion result but then I would have to synchronize on the still-logically-immutable object.